### PR TITLE
Support for password protected servers

### PR DESCRIPTION
**Pull request by mmjr-x**
This PR adds the ability to host and join password protected servers.

It should be good to go, however I will keep this as draft for gathering input :)

Some potential improvements that could be made:
- V - Move the client password input field to the join screen perhaps (or even better have there be a popup to enter a password when joining a password protected server)

When it is done this should close https://github.com/hubastard/nebula/issues/391

### CODE REVIEW

File: NebulaNetwork/Client.cs [hunk 1]
```diff
@@ -66,6 +71,25 @@ public override void Start()
...
+                currentLogOutput(logData, arg2);
+
+                // This method of detecting an authentication failure is super finicky, however there is no other way to do this in the websocket package we are currently using
+                if (logData.Level == LogLevel.Fatal && logData.Message == "Requires the authentication.")
+                {
+                    websocketAuthenticationFailure = true;
+                }
+            };
+
+            if (!string.IsNullOrEmpty(serverPassword))
```
Comment (PhantomGamers) [inline NebulaNetwork/Client.cs:86]: This should probably be IsNullOrWhitespace

File: NebulaPatcher/Patches/Dynamic/UIMainMenu_Patch.cs [hunk 2]
```diff
@@ -185,6 +186,18 @@ private static void AddMultiplayerJoinMenu()
...
             hostIPAddressInput.text = ip;
 
+            Transform passwordField = Object.Instantiate(hostIpField, hostIpField.parent, false);
+            passwordField.localPosition = multiplayerMenu.Find("star-count").localPosition;
+            passwordField.GetComponent<Text>().text = "Password (optional)";
+            passwordInput = passwordField.GetComponentInChildren<InputField>();
+            passwordInput.contentType = InputField.ContentType.Password;
+
+            passwordInput.text = "";
+            if (Config.Options.RememberLastClientPassword && !string.IsNullOrEmpty(Config.Options.LastClientPassword))
```
Comment (PhantomGamers) [inline NebulaPatcher/Patches/Dynamic/UIMainMenu_Patch.cs:196]: IsNullOrWhitespace

File: NebulaPatcher/Patches/Dynamic/VFPreload_Patch.cs [hunk 3]
```diff
@@ -19,7 +19,13 @@ public static void InvokeOnLoadWorkEnded_Postfix()
                     ip = Config.Options.LastIP;
                 }
 
-                UIMainMenu_Patch.JoinGame(ip);
+                string password = "";
+                if (Config.Options.RememberLastClientPassword && !string.IsNullOrEmpty(Config.Options.LastClientPassword))
```
Comment (PhantomGamers) [inline NebulaPatcher/Patches/Dynamic/VFPreload_Patch.cs:23]: IsNullOrWhitespace

File: NebulaNetwork/Server.cs [hunk 4]
```diff
@@ -78,6 +79,20 @@ public override void Start()
             socket = new WebSocketServer(System.Net.IPAddress.IPv6Any, port);
             socket.Log.Level = LogLevel.Debug;
             socket.AllowForwardedRequest = true; // This is required to make the websocket play nice with tunneling services like ngrok
+
+            if (!string.IsNullOrEmpty(Config.Options.ServerPassword))
```
Comment (PhantomGamers) [inline NebulaNetwork/Server.cs:83]: IsNullOrWhitespace

File: NebulaNetwork/Client.cs (hunk 1 above)
Comment (mmjr-x) [reply NebulaNetwork/Client.cs:86]: I'll make the changes

File: NebulaNetwork/Client.cs [hunk 5]
```diff
@@ -262,13 +262,16 @@ private void ClientSocket_OnClose(object sender, CloseEventArgs e)
                 {
                     InGamePopup.AskInput(
                         "Server Requires Password",
-                        "Server is protected. Please enter correct password:",
+                        "Server is protected. Please enter the correct password:",
                         InputField.ContentType.Password,
                         Config.Options.ClientPassword,
                         (password) =>
                         {
+                            Multiplayer.ShouldReturnToJoinMenu = false;
```
[the remark this answers is not in the document]
Comment (mmjr-x) [reply NebulaNetwork/Client.cs:277]: @starfi5h would still like your view on this? :)

**Review by starfi5h — APPROVED** [review]

**Pull request merged**